[WIP] API proposal for handling incoming sessions/links

alanconway opened the pull request:
Initial implementation REVIEW ONLY - basic server example

example_server_test.go: very simple one-queue broker, tests basic send/receive.

- Lots of unfinished stuff: grep for FIXME to see known holes.
- Needs much more testing, performance profiling etc.

Depends on the following pending PRs, must be re-based after they merge:

- #204  Improved debug logging.
- #203  Make conn.Close() wait till shutdown is complete.
- #201  Rename Client to Conn
- #198  Fix SEGV when using dynamic addresses.

grs commented on conn.go:
```diff
@@ -156,6 +156,36 @@ func ConnContainerID(id string) ConnOption {
+// SASL negotiation. Once that is done, AMQP is perfectly symmetric.
+
+// ConnSession sets a callback for incoming session begin.
+// To reject the session, call Detach().
```
Why Detach() and not e.g. End() to match protocol (or Close() as a more common generic shutdown action).

alanconway replied:
Agreed. 

Renamed to CloseErr(). The API already offers a uniform Close() for orderly close so CloseErr() fits the existing style.

Made some other changes too: I replaced the multiple handler options with a single handler interface. This fixes a pitfall in electron's API where you want to accept Receivers but forget that you also have to accept Sessions or you'll never get a Receiver. Now you have to implement a Handler that explicitly says what to do for all 3 cases.

The new Handler gets an IncomingSender (etc.) object containing the begin/attach performative data *not* a usable *Sender object. It can call IncomingSender.Accept() to accept and get a usable *Sender, or return amqp.Error to reject. 

I thought I could blur that distinction for simplicity but on reflection I can't, otherwise it's too easy for applications to attempt to Send() on a Sender that isn't open yet, with hilarious consequences (not). This would create a nasty race condition  when handlers spin up goroutines to  handle endpoints concurrently, unless we have a clear point at which we say "ok now you can use it" - which is what the Accept() call is for.

csibbitt commented:
I recently moved one of our test tools over from electron to this library and it almost doubled our performance (https://github.com/redhat-service-assurance/telemetry-bench/pull/10). I went looking for the server-side to try out and discovered this! The API described here matches roughly what I was expecting to find, so I'm happy to be an early tester and provide additional PR feedback as the implementation materializes. Thanks!

chirino commented on client.go:
```diff
@@ -89,6 +93,11 @@ func Dial(addr string, opts ...ConnOption) (*Client, error) {
 }
 
+// Accept accepts an AMQP server connection.
+func Accept(l net.Listener, opts ...ConnOption) (*Client, error) {
```
Feeling we could just let the user handle doing the a tcp or tls Listener and accept the net.Conn.  Then the user uses the ConnServer() /w the existing New func. That would keep the API a bit cleaner I think.

alanconway replied:
Thats how it works now, check the latest.

alanconway commented:
On Thu, Oct 17, 2019 at 1:39 PM Hiram Chirino <notifications@github.com>
wrote:

> *@chirino* commented on this pull request.
> ------------------------------
>
> In client.go
> <https://github.com/vcabbage/amqp/pull/178#discussion_r336137313>:
>
> > @@ -89,6 +93,11 @@ func Dial(addr string, opts ...ConnOption) (*Client, error) {
>  	return &Client{conn: c}, err
>  }
>
> +// Accept accepts an AMQP server connection.
> +func Accept(l net.Listener, opts ...ConnOption) (*Client, error) {
>
> Feeling we could just let the user handle doing the a tcp or tls Listener
> and accept the net.Conn. Then the user uses the ConnServer() /w the
> existing New func. That would keep the API a bit cleaner I think.
>
We could, but the same can be said of Dial() - both are just convenience
wrappers around New(). I'm inclined to have Accept() for symmertry with
Dial() but I don't feel strongly either way.

vcabbage commented:
Thanks for working on this!

I feel like there's a bit of an mismatch between the server API and the existing client API, largely because I was only thinking of the client perspective when I originally wrote it.

Couple notes on the proposed design:
* I don't want to directly expose the internal `perform...` types. Creating new exported types that have some or all of their fields is fine.
* I'm not sure accepting a `Handler` is a good fit for AMQP and is somewhat at odds with the client API. Perhaps it could work with some refinement though.

Here's what I came up with, which ended up being very similar to your design but without the `Handler`. It didn't start out that way; I didn't fully appreciate some aspects of your design until doing it myself and I'm not convinced this fits better than yours :smile:. I think removing the `Handler` makes it both more complex and more flexible for the user, which may be appropriate since this is intended to be a relatively low-level library. Having never implemented an AMQP server it's possible my thinking is misguided, any feedback you have is appreciated.


``` go
// Rename `Client` to `Conn`. Use type alias to deprecate `Client`
// without introducing an immediate breaking change.

func ConnAllowIncoming() ConnOption {} // For client connections, default for Conns via IncomingConn.

func NewListener(c net.Conn) Listener {}
func Listen(addr string, ListenOptions...) (Listener, error) {} // Maybe defer on this.

type Listener struct {}
func (Listener) NextIncoming() (IncomingConn, error) {}

type IncomingConn struct {
    // ...relevant open fields...
}
func (IncomingConn) Accept(...ConnOptions) (*Conn, error) {}
func (IncomingConn) Close(*Error) error {}

func (*Conn) NextIncoming() (IncomingSession, error) {}

type IncomingSession struct {
    // ...relevant begin fields...
}
func (IncomingSession) Accept(...SessionOptions) (*Session, error) {}
func (IncomingSession) Close(*Error) error {}

// Must be type asserted to accept the Sender/Receiver.
type IncomingLink interface {
    Address() string
    Close(*Error) error
}

type IncomingSender struct {
    // ...relevant attach fields...
}
func (IncomingSender) Accept(...LinkOptions) (*Sender, error) {}
func (IncomingSender) Address() error {}
func (IncomingSender) Close(*Error) {}

type IncomingReceiver struct {
    // ...relevant attach fields...
}
func (IncomingReceiver) Accept(...LinkOptions) (*Receiver, error) {}
func (IncomingReceiver) Address() error {}
func (IncomingReceiver) Close(*Error) {}
```

alanconway commented:
> * I don't want to directly expose the internal `perform...` types. Creating new exported types that have some or all of their fields is fine.

+1 that was laziness on my part. Electron has IncomingSender/Session/Receiver much like you have painted here.

> * I'm not sure accepting a `Handler` is a good fit for AMQP and is somewhat at odds with the client API. Perhaps it could work with some refinement though.

> Here's what I came up with, which ended up being very similar to your design but \without the `Handler`. It didn't start out that way; I didn't fully appreciate some aspects of your design until doing it myself and I'm not convinced this fits better than yours . I think removing the `Handler` makes it both more complex and more flexible for the user, which may be appropriate since this is intended to be a relatively low-level library. Having never implemented an AMQP server it's possible my thinking is misguided, any feedback you have is appreciated.
>

This is very close to what electron does, actually the Handler idea arose from a pitfall of electron where it's easy to write code to accept links but forget that you also have to accept the parent sessions. Your approach makes the session/link relationship explicit so that's not an issue.

I like your approach - it is explicit, follows the AMQP model, and sticks to simple blocking calls - no channels or callbacks. IMO it is logically/runtime equivalent to the Handler approach but more "obvious" which is a good thing.

It does involve some boilerplate - to write a Send or Receive loop you need to write 2 other loops to accept sessions and links. However that can be easily solved by implementing something like the Handler as a convenience on top of your API so we can have our cake and eat it - clarity and user convenience. It might be that the extra loops aren't that big a deal and your API is fine as it stands.

So +1 to this design.

alanconway commented:
Doh wrote all that but missed the main event:

Accepting incoming endpoints *must* be serialized: you basically have no choice but to stop reading the connection until a decision is made. Otherwise you can get incoming opens for links belonging to sessions that are not yet locally open and other catastrophes. AMQP allows pipelining opens, so the remote can open a session and start opening links *before* it's received a session open. 

So that means it's critical that opens are handled promptly - dedicated goroutine that doesn't do a lot of extra work etc. If you declare "AllowsIncoming" you *must* have goroutines sitting in all the Next...() calls at all times. Of course you can do that with your API, but its a potential pitfall if someone understand it fully - e.g. a client that only expects one incoming session/link might processes them and then stop calling Next() - which would cause the client to mysteriously hang if the remote tries to open an unexpected link.

The benefit of a single Handler is it forces the user to say what to do with any incoming endpoint, and the Client goroutines can always make the call in the proper sequence. Your API will have the Client blocking on the assumption that the relevant Next() function will be called. Both can go wrong if the user misocdes things but the Handler approach is more likely to report a useful error whereas the Next() approach is likely to hang mysteriously.

Not ruling out your approach, it still has a lot of things I like and there are pitfalls either way (Handler is called in unknown goroutine context) but that's the biggest usability trade-off I think.

chirino commented:
I do tend to like this API too.  To assist the bad developer case, a timeout could be used to log warning  to let the developer know that there was a pipelined open that is stalling the connection due to him not processing it.

alanconway commented:
We can have-cake-and-eat: We use Kales NextIncoming() API. For total control you can start your own goroutines and use it directly - but do it right or things will hang.

For common simple cases, we provide this optional convenience function:
```
// HandleIncoming starts goroutines as needed to loop on NextIncoming() for Listener, 
// Connection and Session, and calls the relevant callback function. 
// If a callback is nil, the default behavior is to call incoming.Accept(). 
// You should not call NextIncoming() elsewhere in your code if you use this.
HandleIncoming(Listener, func(IncomingConn), func(IncomingSession), func(IncomingLink))
```

Now common cases like "I only care about Receivers, I want a goroutine for each"  become:
```
HandleIncoming(listener, nil, nil, func(link IncomingLink) { 
  switch link := link.(type) { 
    case Receiver: go myHandleReceiver(link);
    case Sender: // ignore, or do link.CloseErr("I don't want senders")
  }
}
```

That is the neatest yet, and we still have a clean blocking-call NextIncoming() API underneath. 
We'll provide a correct, general-purpose HandleIncoming, but people who want to craft their own concurrency model can do exactly what they want - at their own risk.

alanconway commented:
Separate issue: we need to be clear about the distinction between ConnServer and ConnAllowIncoming:

Any connection, client or server, can have ConnAllowIncoming(). If ConnAllowIncoming is set, the user MUST call HandleIncoming() or start NextIncoming() goroutines. (That was the idea behind passing a Handler in the option - if you ask for incomings you must say what to do with them. I'm sold on the NextIncoming() API so I guess that's just one for the docs now.)

A server connection is one from net.Listener.Accept(). AMQP makes no difference between client and server connections, but SASL does, so just for the SASL handshake you have to know which you are. After that everything is symmetric.

chirino commented:
One this api change goes in calling things a client might seem weird. But renaming that as this point might be too large a change.

chirino commented:
should we pass in a Context into methods that start go routines? That way we can check to see if the caller wants to cancel it.

alanconway commented:
> One this api change goes in calling things a client might seem weird. But renaming that as this point might be too large a change.

+1 although Kale's comment implied he was thinking about the Client->Conn change. It'll work either way but Conn would be clearer.

> should we pass in a Context into methods that start go routines? That way we can check to see if the caller wants to cancel it.

I would say no - the lifetime of these goroutines is directly tied to the lifetime of the Client, there's no reason for them to exist after the Client closes, and its an error (could hang the process) to cancel them before.

leifmadsen commented:
> We can have-cake-and-eat: We use Kales NextIncoming() API. For total control you can start your own goroutines and use it directly - but do it right or things will hang.
> 
> For common simple cases, we provide this optional convenience function:
> 
> ```
> // HandleIncoming starts goroutines as needed to loop on NextIncoming() for Listener, 
> // Connection and Session, and calls the relevant callback function. 
> // If a callback is nil, the default behavior is to call incoming.Accept(). 
> // You should not call NextIncoming() elsewhere in your code if you use this.
> HandleIncoming(Listener, func(IncomingConn), func(IncomingSession), func(IncomingLink))
> ```
> 
> Now common cases like "I only care about Receivers, I want a goroutine for each" become:
> 
> ```
> HandleIncoming(listener, nil, nil, func(link IncomingLink) { 
>   switch link := link.(type) { 
>     case Receiver: go myHandleReceiver(link);
>     case Sender: // ignore, or do link.CloseErr("I don't want senders")
>   }
> }
> ```
> 
> That is the neatest yet, and we still have a clean blocking-call NextIncoming() API underneath.
> We'll provide a correct, general-purpose HandleIncoming, but people who want to craft their own concurrency model can do exactly what they want - at their own risk.

I have been following along, and this exact thought came in my head. Why preclude both? If the API is built with primitives, and then provide a usable implementation of those primitives that can be safely used that'd be awesome. Then the primitives can be exposed if someone needs to build something differently.

It would even allow for the addition of primitives and some sort of internal versioning of the implementation too in case you needed to extend or change the API.

alanconway commented:
I've updated the PR based on this discussion, I quite like it. The simple server example is fairly neat.
I made some minor changes as I went, they'e all marked with FIXME comments for easy review.

vcabbage commented:
I'm hesitant about the `HandleIncoming` helper. If a callback is `nil`, what options will be used when accepting? How do the `IncomingConn` and `IncomingSession` handlers work, does the accepted conn/session need to be returned from the callback?

I understand the desire for a simplified API and I think it's worthwhile, but I'm not sure trying to do it right away is the best course. I'm leaning towards having the initial server support be only the basics. Helpers can always be added after that's merged.

* The latest updates have `HandleIncoming(*Client, func(*IncomingSession), func(*IncomingSender), func(*IncomingReceiver))` instead of `HandleIncoming(Listener, func(IncomingConn), func(IncomingSession), func(IncomingLink))`. Is that intentional?
* I'm partial towards `Accept()` rather than `New()`. To me it describes what is being the done; the incoming request for a conn/session/link is being accepted.
* As discussed in previous comments, yes, I do think `Client` should be renamed to `Conn`. Using a type alias will allow there to be a deprecation period.
```go
type Client = Conn
```
* I failed to provide details regarding the `ConnAllowIncoming()` option. My intention was that if clients don't set this option incoming sessions will be automatically rejected, making it easier for the typical client cases.

alanconway commented:
On Sun, Oct 27, 2019 at 5:40 PM Kale Blankenship <notifications@github.com>
wrote:

> I'm hesitant about the HandleIncoming helper.
>
+1 - the primary API is the thing to do first. We can reconsider helpers
when it is in place.

> If a callback is nil, what options will be used when accepting?
>
Exactly equivalent to calling Incoming*.Accept() with no options.
The library has to compute defaults regardless of the API; they're pretty
obvious as I recall, either empty values or copied from incoming.

> How do the IncomingConn and IncomingSession handlers work, does the
> accepted conn/session need to be returned from the callback?
>
Good point - and evidence that we should finish the primary API first :)

>
>    - The latest updates have HandleIncoming(*Client,
>    func(*IncomingSession), func(*IncomingSender), func(*IncomingReceiver))
>    instead of HandleIncoming(Listener, func(IncomingConn),
>    func(IncomingSession), func(IncomingLink)). Is that intentional?
>
> I tend to think of amqp.Conn as the "topmost" object and "listeners" as
something outside the AMQP API.
I prefer this:
    inConn, err := amqp.Accept(l)
To this:
    al := amqp.NewListener(l) // Not really adding anything
    inConn, err := al.Accept()

Side issue: electron has a Container with Dial/Accept/New methods. Its only
role is to stamp a consistent container-id on all the connections, but
that's otherwise a bit of a pain to do with ConnContainerId() for every
single connection. Generally nobody cares about the client container-id,
but for servers it can be relevant that the server uses it's container-id
consistently.

>
>    - I'm partial towards Accept() rather than New(). To me it describes
>    what is being the done; the incoming request for a conn/session/link is
>    being accepted.
>
> Accept was my first instinct also (that's what electron uses) and I don't
have any better ideas so I'm OK with it.

>
>    - As discussed in previous comments, yes, I do think Client should be
>    renamed to Conn. Using a type alias will allow there to be a
>    deprecation period.
>
> type Client = Conn
>
> +1

>
>    - I failed to provide details regarding the ConnAllowIncoming()
>    option. My intention was that if clients don't set this option incoming
>    sessions will be automatically rejected, making it easier for the typical
>    client cases.
>
> +1

leifmadsen commented:
> I understand the desire for a simplified API and I think it's worthwhile, but I'm not sure trying to do it right away is the best course. I'm leaning towards having the initial server support be only the basics. Helpers can always be added after that's merged.

Makes enough sense. Once the primitives are in place, folks can create their own helpers, and eventually it'll become obvious what is missing, what needs to be adjusted, and at that point there should be enough examples around to provide a patch with a helper.

Initially the focus can just be on example documentation using the primitives, then things should eventually snap into place.

alanconway committed:
Rename Client to Conn

Rename Client to Conn withj backwards-compatible Client alias.

Signed-off-by: Alan Conway <aconway@redhat.com>

alanconway commented:
I've consolidated all comments in the new PR. I renamed Client to Conn, the new stuff is in incoming.go. Please see example_server_test.go which is a complete exmample. Things to note:

- There's no amqp.Listener - the starting point for the API is always Conn (from Dial() or New()) or IncomingConn (from NewIncoming()).
- There's no ConnServer option - NewIncoming() creates an IncomingConn for a server connection. It doesn't *have* to be a Listen.Accept() connection, e.g. you could do AMQP over a net.Pipe() - but one end would have to act as server for SASL.
- ConnAllowIncoming() is never the default, the user must always set it - this could be debatable, but I feel it emphasizes that all amqp.Conn are the same as far as use or not of  NextIncoming(). Also if we default it we have to provide a way to turn it off in the (unusual but not impossible) case of a server that initiates all sessions/links itself.

Overall I am pretty happy with the example  - some boilerplate, but it's not too bad.

alanconway commented:
I've started to implement IncomingConn and IncomingSession - this is not complete or working yet but I'd appreciate any comments on the approach and style, and whether it fits with the rest of the project.

vcabbage reviewed (commented):
Looks good so far :+1: I left a couple nit-picky comments :slightly_smiling_face:

alanconway commented on client.go:
```diff
@@ -135,26 +153,40 @@ func (c *Client) NewSession(opts ...SessionOption) (*Session, error) {
+func (c *Conn) NextIncoming() (*IncomingSession, error) {
+	if c.conn.incomingSession == nil {
+		// This is a logic error, not a runtime error.
+		panic("Conn.NextIncoming requires the ConnAllowIncoming option")
```
Done. Although to counter-nit: https://github.com/vcabbage/amqp/blob/master/client.go#L781 ;)

vcabbage replied:
Fair point :) That one is supposed to be impossible, if someone is able to trigger it it's because of a bug in the library.

alanconway replied:
I get it, I just couldn't resist.

alanconway commented on conn.go:
```diff
@@ -370,16 +404,31 @@ func (c *conn) mux() {
+					// Note this blocks the conn mux until the application calls NextIncoming()
+					// FIXME(alanconway) review blocking & safety.
+					// XXX
+					c.incomingSession <- ic
```
how about  a `switch {` instead of the `} else if {`?

alanconway commented on conn.go:
```diff
@@ -320,21 +329,46 @@ func (c *conn) getErr() error {
+		byChannel:       make(map[uint16]*Session),
+		byRemoteChannel: make(map[uint16]*Session),
+		conn:            c,
+	}
```
Yes, I liked your original encapsulation but it got too repetitive when I added the server stuff.
The conn is used in `advance()`  but I moved it out of the struct and made it a parameter.

alanconway commented on incoming.go:
```diff
@@ -0,0 +1,224 @@
+	if ic.conn, err = newConn(nc); err != nil {
+		return nil, err
+	}
+	if err = ic.conn.start(ic.negotiateProto); err != nil {
```
Done. My fingers may rebel, apologies in advance if I break the rule again.
Do you have a style guide listing your preferences?

alanconway committed:
Fix SEGV when using dynamic addresses.

Fixes  #197

Signed-off-by: Alan Conway <aconway@redhat.com>

alanconway committed:
Export types that need to be exposed in Incoming objects.

Symbol, Symbols (was multiSymbol), Source, Target, FilterSet (was filter),
DescribedType

Signed-off-by: Alan Conway <aconway@redhat.com>

alanconway committed:
WIP: #178 REVIEW ONLY - basic server example

example_server_test.go: very simple one-queue broker, tests basic send/receive.

- Lots of unfinished stuff: grep for FIXME to see known holes.
- Needs much more testing, performance profiling etc.

Depends on the following pending PRs, must be re-based after they merge:

- #204  Improved debug logging.
- #203  Make conn.Close() wait till shutdown is complete.
- #201  Rename Client to Conn
- #198  Fix SEGV when using dynamic addresses.

Signed-off-by: Alan Conway <aconway@redhat.com>

alanconway committed:
Server and Incoming API with example - no implementation yet.

Start with example_server_test.go, then see incoming.go.

Signed-off-by: Alan Conway <aconway@redhat.com>

alanconway committed:
Improved debug logging.

Added connection id (pointer) and channel for debugging multi-connection/session programs.

Centralized scattered debug logic to fix missing/duplicate log entries.

Removed (Session) and (Link) decorations - these can be deduced from the frame type.

Don't print null fields in log output - reduce distracting clutter.

Signed-off-by: Alan Conway <aconway@redhat.com>

alanconway committed:
Make conn.Close() wait till shutdown is complete.

This showed up with the improved logging: log messages for the Close frames of one
test were printed after another test had started.

Should be deadlock safe, since we don't call user callbacks in the mux or tx/rx goroutines.

Signed-off-by: Alan Conway <aconway@redhat.com>

vcabbage commented:
I really appreciate the all the contributions both in code and discussion here. Unfortunately I'm not going to be maintaining this repo anymore, details in #205.
